### ISSUE TITLE
nx migrate for Angular material v17 takes a lot of time then revert it back to v16

### ISSUE THREAD

**Issue by robertIsaac** [nrwl/nx#20301]
### Current Behavior

when running `nx migrate next` which uses `17.2.0-beta.0` initially upgrade @angular/material to 17.0.1
after running the `pnpm exec nx migrate --run-migrations --create-commits` command I see the last commit which says `chore: [nx migration] migration-v17` is changing it back to `^16.2.0` 

### Expected Behavior

to correctly run the migration to v17 which includes changes to `styles.scss`

### GitHub Repo

I can give you access if needed

### Steps to Reproduce

in Angular v16 project with @angular/material 
1. run `nx migrate next`
2. run `pnpm exec nx migrate --run-migrations --create-commits`

### Nx Report

```shell
Node   : 18.16.0
   OS     : win32-x64
   pnpm   : 8.6.0

   nx                 : 17.2.0-beta.0
   @nx/js             : 17.2.0-beta.0
   @nx/jest           : 17.2.0-beta.0
   @nx/linter         : 17.2.0-beta.0
   @nx/eslint         : 17.2.0-beta.0
   @nx/workspace      : 17.2.0-beta.0
   @nx/angular        : 17.2.0-beta.0
   @nx/cypress        : 17.2.0-beta.0
   @nx/devkit         : 17.2.0-beta.0
   @nx/eslint-plugin  : 17.2.0-beta.0
   @nx/nest           : 17.2.0-beta.0
   @nx/node           : 17.2.0-beta.0
   @nrwl/tao          : 17.2.0-beta.0
   @nx/web            : 17.2.0-beta.0
   @nx/webpack        : 17.2.0-beta.0
   typescript         : 5.2.2
   ---------------------------------------
   Community plugins:
   ng-mocks : 14.11.0
```

### Operating System

- [ ] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

1. running `nx migrate @angular/material@latest` after then `pnpm exec nx migrate --run-migrations --create-commits` doesn't help, it repeats the behavior
2. running ng update on the same project after creating custom `angular.json` works fine
3. I hope you don't hate me for reporting many bugs

**Comment by leosvelperez** [comment]
Thanks for reporting this!

No worries about reporting issues, that only helps to make the product better, so keep them coming 👍🏻 

There are a couple of things at play here, none of them caused by Nx, though they don't happen in an Angular CLI workspace because of a difference on how package updates are applied. Still, there's a fundamental issue in the Angular Material migration that cause it to be very slow as you noticed.

tldr; this is only an issue with `pnpm` (`npm` and `yarn` are fine) and it's caused by the Angular Material migration visiting and processing files within `node_modules`.

## In-depth explanation

The Angular Material v17 migration:

- Has an initial validation checking whether the workspace is using legacy imports (`@angular/material/legacy-`)
- To validate that, the migration visits every single file in the workspace, including ignored files and process any files ending with `.ts` (including `d.ts` which you would encounter in `node_modules`). That means, among others, all files inside `node_modules` will be processed. This is why is taking that much time to process that migration. The migration should not visit ignored files.

On top of that, when using `pnpm`, you can end up with multiple versions of `@angular/material` in the `node_modules/.pnpm` directory. The 16.x.x versions (what you have before migrating) will have `.d.ts` files containing legacy imports. That's why the migration throws and reverts back to v16, stating that it cannot update the project to v17.

Below is an example failure logs where those files in `node_modules` can be seen reported:

<details>

<summary>Failure logs</summary>

```shell
Cannot update to Angular Material v17 because the project is using the legacy Material components
    that have been deleted. While Angular Material v16 is compatible with Angular v17, it is recommended
    to switch away from the legacy components as soon as possible because they no longer receive bug fixes,
    accessibility improvements and new features.
    
    Read more about migrating away from legacy components: https://material.angular.io/guide/mdc-migration
    
    Files in the project using legacy Material components:
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-autocomplete/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-autocomplete/testing/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-chips/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-dialog/testing/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-form-field/testing/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-input/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-list/testing/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-paginator/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-paginator/testing/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-progress-spinner/testing/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-select/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-select/testing/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-snack-bar/index.d.ts
     - /node_modules/.pnpm/@angular+material@16.2.12_@angular+animations@16.2.12_@angular+cdk@16.2.12_@angular+common@16_2es3biwgpoqfqgffdq3u6n2fgu/node_modules/@angular/material/legacy-table/index.d.ts
```

</details>

Now, as you mentioned, the failure doesn't happen in an Angular CLI workspace (though the it still visits all files in the workspace, including the ignored ones). This is because the `ng update` command, before installing the new package versions, it wipes out `node_modules`. So, but the time the migration runs, the previous version which would contain files with legacy imports is not present in the `node_modules/.pnpm` directory. Only the new version (v17) is installed and that one doesn't contain any `.d.ts` files with legacy imports.

While the `nx migrate` command could potentially be changed to wipe out `node_modules` before installing the new package versions, the real issue here is that the Angular CLI migration should not process those files in the first place. This can be reported in the Angular CLI repo. Even though in an Angular CLI repo the migration succeeds, it still has a degraded performance due to visiting ignored files. 

I'm going to close this issue since the root of the problem is in the Angular Material migration. We'll discuss internally whether it's beneficial to wipe out `node_modules` before installing new package versions (it could have a negative perf impact due to additional installation time).

**Comment by robertIsaac** [comment]
> Thank @leosvelperez for such a detailed report! Trying to migrate and this is very helpful info

@AdditionAddict FYI it's fixed in the latest version
